irmaseal-core/src/meta.rs

```diff
@@ -0,0 +1,95 @@
+use super::stream::*;
+use crate::Identity;
+use super::{Error};
+use base64;
+use arrayvec::{ArrayString, ArrayVec};
+use serde::{Deserialize, Serialize, Deserializer, Serializer};
+
+/// The version of the encrypted data
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum Version {
+    V1_0,
+}
+
+/// Metadata which contains the version
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct Metadata {
+    pub version: Version,
+    pub media_type: ArrayString<[u8; 255]>,
+    pub media_metadata: serde_json::Value,
+    #[serde(serialize_with = "to_base64_ciphertext", deserialize_with = "from_base64_ciphertext")]
+    pub ciphertext: ArrayVec<[u8; CIPHERTEXT_SIZE]>,
+    #[serde(serialize_with = "to_base64_iv", deserialize_with = "from_base64_iv")]
+    pub iv: ArrayVec<[u8; IVSIZE]>,
+    pub identity: Identity,
+}
+
+// Can't have natural at the type level(yet). so just use to_base64 twice with different sizes.
+fn to_base64_ciphertext<S>(ciphertext: &ArrayVec<[u8; CIPHERTEXT_SIZE]>, serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
If you use a fixed size array for `iv` and `ciphertext`, and these fixed size arrays all have a `as_slice` method. Maybe serde understands this such that you can make one general function for both?

Reply:
Once we have const generics it should be possible but alas not yet.